dpnp/dpnp_algo/dpnp_algo.pyx

```diff
@@ -235,8 +236,19 @@ cpdef dparray dpnp_matmul(dparray in_array1, dparray in_array2):
     cdef DPNPFuncData kernel_data = get_dpnp_function_ptr(DPNP_FN_MATMUL, param1_type, param2_type)
 
     result_type = dpnp_DPNPFuncType_to_dtype(< size_t > kernel_data.return_type)
-    # ceate result array with type given by FPTR data
-    cdef dparray result = dparray(shape_result, dtype=result_type)
+
+    cdef dparray result
+
+    if out is not None:
+        if out.dtype != result_type:
+            raise TypeError("Inappropriate type of ``out`` variable")
+        elif out.shape != shape_result:
+            raise ValueError("Inappropriate shape of ``out`` variable")
```

Comment:
I would use `checker_throw_value_error` in these cases, e.g.:
`checker_throw_value_error('matmul', 'out.dtype', out.dtype, result_type)`

Reply:
You could replace `elif` with `if` and remove line with `else`.

Reply:
No error raising on cython layer.

Reply:
We don't know `result_type` in python level.